specs/core/1_shard-data-chains.md

```diff
@@ -0,0 +1,122 @@
+
+```python
+def calc_block_maxbytes(state):
+    max_grains = MAX_SHARD_BLOCK_SIZE // CHUNK_SIZE
+    validators_at_max_committees = SHARD_COUNT * TARGET_COMMITTEE_SIZE
+    grains = min(len(get_active_validator_indices(state.validators)) * max_grains // validators_at_max_committees,
+                 max_grains)
+    return CHUNK_SIZE * grains
+```
+
+* Verify that `len(block_body) == calc_block_maxbytes(state)`
+* Define `filler_bytes = next_power_of_2(len(block_body)) - len(block_body)`. Compute a simple binary Merkle tree of `block_body + bytes([0] * filler_bytes)` and verify that the root equals the `data_root` in the header.
+
+### Verifying a crosslink
+
+A node should sign a crosslink only if the following conditions hold. **If a node has the capability to perform the required level of verification, it should NOT follow chains on which a crosslink for which these conditions do NOT hold has been included, or a sufficient number of signatures have been included that during the next state recalculation, a crosslink will be registered.**
+
+First, the conditions must recursively apply to the crosslink referenced in `last_crosslink_hash` for the same shard (unless `last_crosslink_hash` equals zero, in which case we are at the genesis).
+
+Second, we verify the `shard_block_combined_data_root`. Let `B[0]` be the first block _after_ the last crosslink and `B[n-1]` be the block directly referenced by the `shard_block_hash`. Let `bodies[0] .... bodies[n-1]` be their bodies and `roots[0] ... roots[n-1]` the data roots. Define `compute_merkle_root` be a simple Merkle root calculating function that takes as input a list of objects, where the list's length must be an exact power of two. Let `state[0] ... state[n-1]` be the beacon chain states at those times, and `depths[0] ... depths[n-1]` be equal to `log2(next_power_of_2(calc_block_maxbytes(state[i]) // CHUNK_SIZE))` (ie. the expected depth of the i'th data tree).
+
+```python
+def get_zeroroot_at_depth(n):
+    o = b'\x00' * CHUNK_SIZE
+    for i in range(n):
+        o = hash(o + o)
+    return o
+
+def mk_combined_data_root(depths, roots):
+    default_value = get_zeroroot_at_depth(max(depths))
+    data = [default_value for _ in range(next_power_of_2(len(roots)))]
+    for i, (depth, root) in enumerate(zip(depths, roots)):
+        value = root
+        for j in range(depth, max(depths)):
+            value = hash(value, get_zeroroot_at_depth(depth + j))
+        data[i] = value
+    return compute_merkle_root(data)
+```
+
+Essentially, this outputs the root of a tree of the data roots, with the data roots all adjusted to have the same height if needed. The tree can also be viewed as a tree of all of the underlying data concatenated together, appropriately padded. Here is an equivalent definition that uses bodies instead of roots [TODO: check equivalence]:
```

Comment:
```suggestion
This outputs the root of a tree of the data roots, with the data roots all adjusted to have the same height if needed. The tree can also be viewed as a tree of all of the underlying data concatenated together, appropriately padded. Here is an equivalent definition that uses bodies instead of roots [TODO: check equivalence]:
```

No need for "essentially"